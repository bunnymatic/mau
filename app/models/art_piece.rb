class ArtPiece < ApplicationRecord

  belongs_to :artist

  has_many :art_pieces_tags
  has_many :tags, :through => :art_pieces_tags, :source => :art_piece_tag

  belongs_to :medium

  include TagsHelper

  has_attached_file :photo, styles: MauImage::Paperclip::STANDARD_STYLES, default_url: ''
  validates_attachment_presence :photo
  validates_attachment_content_type :photo,
                                    content_type: ["image/jpg", "image/jpeg", "image/png", "image/gif"],
                                    message: "Only JPEG, PNG, and GIF images are allowed"
  validates :artist_id, presence: true
  include Elasticsearch::Model

  after_commit :add_to_search_index, on: :create
  after_commit :refresh_in_search_index, on: :update
  after_commit :remove_from_search_index, on: :destroy

  def add_to_search_index
    Search::Indexer.index(self)
  end

  def refresh_in_search_index
    Search::Indexer.reindex(self)
  rescue Elasticsearch::Transport::Transport::Errors::NotFound
    add_to_search_index
  end

  def remove_from_search_index
    Search::Indexer.remove(self)
  end

  self.__elasticsearch__.client = Search::EsClient.root_es_client

  settings(analysis: Search::Indexer::ANALYZERS_TOKENIZERS, index: { number_of_shards: 2}) do
    mappings(_all: {analyzer: :mau_snowball_analyzer}) do
      indexes :title, analyzer: :mau_snowball_analyzer
      indexes :year
      indexes :medium, analyzer: :mau_snowball_analyzer
      indexes :artist_name, analyzer: :mau_snowball_analyzer
      indexes :studio_name, analyzer: :mau_snowball_analyzer
      indexes :tags, analyzer: :mau_snowball_analyzer
    end
  end

  before_destroy :remove_images
  after_destroy :clear_tags_and_favorites
  after_save :remove_old_art
  after_save :clear_caches

  validates_presence_of     :title
  validates_length_of       :title,    :within => 2..80

  def as_indexed_json(opts={})
    return {} unless artist && artist.active?

    idxd = as_json(only: [:title, :year])
    extras = {}
    extras["medium"] = medium.try(:name)
    extras["tags"] = tags.map(&:name).join(", ")
    extras["images"] = image_paths
    # guard against bad data
    extras["artist_name"] = artist.full_name
    extras["studio_name"] = artist.studio.name if artist.studio
    extras["os_participant"] = artist.doing_open_studios?

    idxd["art_piece"].merge! extras
    idxd
  end


  def self.find_random(num = 1)
    owned.limit(num).order("rand()")
  end

  def self.owned
    joins(:artist).where( users: { state: 'active' } )
  end

  def tags
    super.alpha
  end


  def clear_tags_and_favorites
    klassname = self.class.name
    Favorite.where(:favoritable_id => id, :favoritable_type => klassname).compact.map(&:destroy)
  end

  def uniq_tags
    tags.uniq_by(&:name)
  end

  def get_name(escape = false)
    escape ? safe_title : self.title
  end

  def safe_title
    HtmlEncoder.encode(self.title)
  end

  def get_path(size = nil, full_path = false)
    size ||= 'medium'
    artpiece_path = image_paths[size.to_sym]
    (full_path ? full_image_path(artpiece_path) : artpiece_path)
  end

  def image_paths
    @image_paths ||= ArtPieceImage.new(self).paths
  end

  def get_paths
    image_paths
  end

  private
  def clear_caches
    ArtPieceService.clear_cache
    if self.artist && self.artist.id != nil?
      SafeCache.delete(artist.representative_art_cache_key)
    end
  end

  def remove_images
    paths = get_paths.values
    paths.each do |pth|
      pth = File.expand_path( File.join( Rails.root, 'public', pth ) )
      if File.exist? pth
        begin
          result = File.delete pth
          ::Rails.logger.debug("Deleted %s" % pth)
        rescue
          ::Rails.logger.error("Failed to delete image %s [%s]" % [pth, $!])
        end
      end
    end
  end

  def remove_old_art
    # mostly this makes stuff work for testing
    if artist
      max = artist.max_pieces
      cur = artist.art_pieces.length
      del = 0
      while cur && max && (cur > max)
        artist.art_pieces.last.destroy
        cur = cur - 1
        del = del + 1
      end
    end
  end

end
