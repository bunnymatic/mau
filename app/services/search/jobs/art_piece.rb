module Search
  module Jobs
    class ArtPiece < ApplicationJob
      queue_as :es_indexer

      def perform(art_piece_id, method)
        art_piece = ::ArtPiece.find_by(id: art_piece_id)
        unless art_piece
          Rails.logger.warn("Unable to find ArtPiece##{object_id} - skipping es indexing")
          return
        end

        Search::Indexer::ArtPieceSearchService.new(art_piece).send(method)
      end
    end
  end
end
