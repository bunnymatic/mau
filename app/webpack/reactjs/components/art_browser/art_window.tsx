import { backgroundImageStyle } from "@js/services";
import { ArtPiece } from "@models/art_piece.model";
import * as types from "@reactjs/types";
import cx from "classnames";
import React, { FC } from "react";

interface AnnotationProps {
  label: string;
  value?: string | number;
  classes?: types.ClassNames;
}

const Annotation: FC<AnnotationProps> = ({ label, value, classes }) => {
  if (!value) {
    return null;
  }
  return (
    <div className={cx("art-window__annotation", classes)}>
      {Boolean(label) && (
        <span className="art-window__annotation-label">{label}</span>
      )}
      <span className="art-window__annotation-value">{value}</span>
    </div>
  );
};

interface ArtWindowProps {
  art: ArtPiece;
}

export const ArtWindow: FC<ArtWindowProps> = ({ art }) => {
  return (
    <div className="art-window">
      <div className="art-window__image-container">
        <div
          className="art-window__image"
          style={backgroundImageStyle(art.imageUrls.original, {
            backgroundSize: "contain",
            backgroundRepeat: "no-repeat",
          })}
        >
          {Boolean(art.hasSold()) && (
            <span className="art-window__sold">Sold</span>
          )}
        </div>
      </div>
      <div className="art-window__info-container">
        <div className="art-window__info--left">
          <Annotation
            label="Title:"
            value={art.title}
            classes="art-window__title"
          />
          <Annotation
            label="Price:"
            value={art.displayPrice}
            classes={cx("art-window__price", {
              "art-window__price--sold": art.hasSold(),
            })}
          />
        </div>
        <div className="art-window__info--right">
          <Annotation
            label="Dimensions:"
            value={art.dimensions}
            classes="art-window__dimensions"
          />
          <Annotation
            label="Medium:"
            value={art?.medium?.name}
            classes="art-window__medium"
          />
          <Annotation
            label="Year:"
            value={art.year}
            classes="art-window__date"
          />
        </div>
      </div>
    </div>
  );
};
