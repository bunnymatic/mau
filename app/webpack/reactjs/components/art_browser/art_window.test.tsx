import { ArtPiece } from "@models/art_piece.model";
import { jsonApiArtPieceFactory as artPieceFactory } from "@test/factories";
import { render } from "@testing-library/react";
import React from "react";

import { ArtWindow } from "./art_window";

describe("ArtWindow", () => {
  let artPiece;
  beforeEach(() => {
    artPiece = new ArtPiece(artPieceFactory.build());
  });
  it("matches the snapshot", () => {
    const { container } = render(<ArtWindow art={artPiece} />);
    expect(container).toMatchSnapshot();
  });
});
