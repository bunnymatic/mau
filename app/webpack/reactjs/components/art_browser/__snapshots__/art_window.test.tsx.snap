// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtWindow matches the snapshot 1`] = `
<div>
  <div
    class="art-window"
  >
    <div
      class="art-window__image-container"
    >
      <div
        class="art-window__image"
        style="background-image: url(original.png); background-size: contain; background-position: center center; background-repeat: no-repeat;"
      />
    </div>
    <div
      class="art-window__info-container"
    >
      <div
        class="art-window__info--left"
      >
        <div
          class="art-window__annotation art-window__title"
        >
          <span
            class="art-window__annotation-label"
          >
            Title:
          </span>
          <span
            class="art-window__annotation-value"
          >
            the title goes here
          </span>
        </div>
        <div
          class="art-window__annotation art-window__dimensions"
        >
          <span
            class="art-window__annotation-label"
          >
            Dimensions:
          </span>
          <span
            class="art-window__annotation-value"
          >
            10x 20
          </span>
        </div>
      </div>
      <div
        class="art-window__info--right"
      >
        <div
          class="art-window__annotation art-window__price art-window__price--sold"
        >
          <span
            class="art-window__annotation-label"
          >
            Price:
          </span>
          <span
            class="art-window__annotation-value"
          >
            $123
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
